enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,205 @@
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the CHNSO has started.
+* CHNSO can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the CHNSO should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address.
+
+#### Disaster recovery support
+
+TODO: describe the disaster recovery use case in the context of CHNSO. 
+
+#### Self hosted DNS services
+
+The self hosted DNS functionality is composed of pods running CoreDNS-mDNS in each node and a NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+The CHNSO will be responsible only for the deployment of the CoreDNS pods while the adjustment of the resolv.conf file will remain in the MCO, in case hosted DNS services should be disabled the CoreDNS instance should be configured as a forwarding server.
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator where all fields are of type Enable/Disable enum:
+
+
+```yaml
+apiVersion: clusterhostednetservices.operator.io/v1alpha1
+kind: ClusterHostedNetServicesConfig
+metadata:
+  name: clusterhostednetservicesconfig-sample
+spec:
+  dns: 
+    nodesresolution: Enable
+    apiresolution: Enable
+    appsresolution: Enable    
+  loadbalancer:
+    apiintloadbalance: Enable
+    defaultingressha: Enable
+status:
+  apiserverinternalipowner: master-0
+  ingressipowner: worker-1
+```
+
+With this sample manifest the CHNSO should:
+- Create DNS hosted service that will resolve node names, api record and .apps wildcard record.
+- Provide high availability for default ingress.
+- Load balance the requests towards api-int between master nodes.
+
+From the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+#### Managing  the cluster hosted networking services
+
+All the LB, DNS, and VIP components will be enabled by default in case of fresh deployment as today, while post-deployment the admin can control the services by editing the singleton CR.
+
+TODO: detailed description of self-hosted API load balancer migration to an external load balancer.
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack.
```

Comment:
does this even technically work? It would mean to remove the /etc/hosts entry, and it would mean to restart all pods that depend on the old API LB IP.

Reply:
I'll add a section with a detailed description of LB migration procedure.